### DISCUSSION TITLE
Adapters configuration

### DISCUSSION THREAD

**Discussion by zguig52** [strapi-community/plugin-io#33] [Q&A]
Hello,
First, thanks for this great work.
I am trying to configure the plugin to use @socket.io/redis-adapter (cluster mode case).
Official doc is here: https://socket.io/docs/v4/redis-adapter/
What I understood : it is needed to init additionnal objects from the redis-adapter lib and call an extra method once the io object is created (io.adapter(args)):
```
import { createAdapter } from "@socket.io/redis-adapter";
import { createClient } from "redis";

const pubClient = createClient({ url: "redis://localhost:6379" });
const subClient = pubClient.duplicate();

const io = new Server();
io.adapter(createAdapter(pubClient, subClient));

```
From what I saw in your conf file and source code, there is only the option to pass args during the "io = new Server()" init.
Have I missed something or is there an additionnal work required to add this support (like adding another field in options for the adapter type and so on)?
Do you see any way to make this work in the actual code status or shall I try to submit a patch to add support for external adapters?

**Comment by ComfortablyCoding** [comment]
Helo @zguig52 ,

You are correct, this wont work with the current way the code is structured, a new feature/patch would need to be added for this to be possible.

A possible solution to this is to add support for hooks. In this case a postInitialization hook that accepts the io instance after instantiation and allows you to modify it before its attached to the actual strapi instance.

Sample below
```js
// plugins.js
module.exports = ({ env }) => ({
  // ...
  io: {
    enabled: true,
    config: {
      IOServerOptions: {
        cors: { origin: "*", methods: ["GET"] },
      },
      contentTypes: {
        message: "*",
        chat: ["create"],
      },
      hooks: {
        postIOInitialization: (io) => {
          const pubClient = createClient({ url: "redis://localhost:6379" });
          const subClient = pubClient.duplicate();
          io.adapter(createAdapter(pubClient, subClient));
          return io;
        },
      },
    },
  },
  // ...
});
```

What do you think? Any suggestions?

**Reply by zguig52** [reply]
Hi @ComfortablyCoding ,

Thanks for your feedback. In the mean time, I had some deep investigations on what could be done to make this work in your plugin.
I changed a few lines in the code to support the adapter option directly, didn't know about the hook possibility. And it finally worked, after fighting with redis cluster :)

Here is what I did, tell me if this seem good for you and I shall/can submit the corresponding pull request?

CONFIG PART:

In my app package.json, I have added:
```
    "@socket.io/redis-adapter": "7.2.0",
    "ioredis": "5.2.3",
```

In the plugin.js, I have added:
```
const { Cluster } = require("ioredis");
const { createAdapter } = require("@socket.io/redis-adapter");
const pubClient = new Cluster([
	{
		host: "cache01-api.mydomain",
		port: 6379,
	},
	{
		host: "cache02-api.{{ host }}.{{ domain }}",
		port: 6379,
	},
...
], {
  redisOptions: {
    password: "MYSTRONGPASSWORD",
  },
});
const subClient = pubClient.duplicate();
```
==> These parts can be adapted to take in account the adaptor required (here I used the redis one, but this can be changed to whatever adapter is needed).


CODE EVOLUTION PART

Hereafter my changes to the plugin:
- In the plugin.js, I have added 1 option IOServerAdapter:
```
      IOServerOptions :{
...
      },
      IOServerAdapter: createAdapter(pubClient, subClient),
      contentTypes: {
...
```

Afterwhile I have declared this extra param in 
src/server/config/index.js
```
		return {
			IOServerOptions: {},
			IOServerAdapter: null,
			contentTypes: {},
			events: [],
			publicRoleName: 'Public',
		};
```

In the bootstrap.js, I have added this parameter to the IO constructor:
```
	// initialize
	strapi.$io = new IO(normalizedSettings.IOServerOptions, normalizedSettings.IOServerAdapter);
```

And finally I could use it in the initializer (server/structures/IO.js:
```
	constructor(options, adapter) {
		this._socket = new Server(strapi.server.httpServer, options);
		if(adapter){this._socket.adapter(adapter)};
		this._socket.use(handshake);
	}
```

I don't know if this is better than the hook strategy (didn't test it yet, if you want I can)?

**Reply by ComfortablyCoding** [reply]
Sorry for the late reply it's been a busy week.

That is one way to go and I appreciate the thorough info but I still prefer going the hook route as it's a more dynamic approach that only accounting for the adapter.

**Comment by ComfortablyCoding** [comment]
This is now possible in v2 via the [init hook](https://strapi-plugin-io.netlify.app/examples/hooks.html#init)
